### DISCUSSION TITLE
Create a low allocation and faster StringBuilder - Span in Action

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#23] [General]
# Create a low allocation and faster StringBuilder - Span in Action

.NET ships a nice StringBuilder since the dawn of time or at least since the beginning of the framework itself. The intention is simple: If we need to concatenate a lot of strings we can't rely on the + operator as we introduce a lot of unnecessary allocations plus it is slow!
That is where the StringBuilder jumps into the picture.
Why is that? Why is the StringBuilder better? And can we do better? Spoiler: Yes!

https://steven-giesel.com/blogPost/4cada9a7-c462-4133-ad7f-e8b671987896

**Comment by snielsson** [comment]
Just stumbled upon your blog I find it really good and useful. Thanks and keep up the good work !

**Reply by linkdotnet** [reply]
Thank you very much for the kind words! Much appreciated!

**Comment by puschie286** [comment]
There is one thing i’m curious about: you rent some memory and return it right after copying your data to it - but keep a reference to it ? after checking your link to array buffer, this looks totally wrong - and would result in undefined behavior.
Am i wrong or do i miss understand something ?

**Reply by linkdotnet** [reply]
Yeah not that shouldn't be there and as you said that results in weird behavior. The reason it works in a simple console app is that no one else will rent some memory (in this short time span). I will remove that line from the code.

The "real" `ValueStringBuilder` doesn't do it and uses `Dispose` to return the rented array to the pool. See here: https://github.com/linkdotnet/StringBuilder/blob/main/src/LinkDotNet.StringBuilder/ValueStringBuilder.cs#L280

Only the "old" rented array should be returned: https://github.com/linkdotnet/StringBuilder/blob/main/src/LinkDotNet.StringBuilder/ValueStringBuilder.cs#L299

**Reply by puschie286** [reply]
ah yes, thats what i thought how it should looks like
thanks for clarification

**Reply by linkdotnet** [reply]
Thanks for the input again @puschie286 . Very true and valuable!
